src/Learning/MHHistograms/MHDiscretizerMODLHistogram.cpp

```diff
@@ -109,33 +111,50 @@ void MHDiscretizerMODLHistogram::DiscretizeValues(ContinuousVector* cvSourceValu
 			oaResultHistograms.Add(postprocessedOptimizedHistogram);
 		oaResultHistograms.Add(optimizedHistogram);
 
-		// Recherche du meilleur histogramme
-		if (postprocessedOptimizedHistogram != NULL)
-			bestHistogram = postprocessedOptimizedHistogram;
-		else
-			bestHistogram = optimizedHistogram;
-
-		// Cas avec contrainte sur le nombre max d'intervalles
+		// Les histogrammes MODL sont optimaux.
+		// Pour chaque variable numerique, on produit une serie d'histogrammes interpretables, par granularite croissante.
+		// Le probleme est que l'histogramme interpretable le plus fin est parfois tres complexe, de type "herisson",
+		// ce qui peut derouter un utilisateur non expert. Ce probleme se produit dans environ 20% des cas.
+		// Plutot que de montrer par defaut l'histogramme interpretable le plus fin, on propose d'utiliser la
+		// regle de Terell-Scott  (cf. https://en.wikipedia.org/wiki/Histogram), avec un nombre de bins de (2 N)^1/3
+		// Ce choix permet d'avoir par defaut un histogramme a la fois fin et plus simple a interpreter:
+		// - en evitant des choix trop heuristiques de type "regle du coude", ou nombre maximal de "peaks"
+		// - pas trop parcimonieux (comme la regle de Sturges en log(N))
+		// - ne dependant pas des valeurs (comme la regle de Scott, qui utilise l'ecart type)
+		// - tres simple a calculer
+		// - avec une justification theorique, facile a defendre
+		nTerellScottMaxPartNumber = int(ceil(pow(2 * cvSourceValues->GetSize(), 1.0 / 3)));
+		nMaxPartNumber = nTerellScottMaxPartNumber;
+
+		// La regle de Terell-Scott est pertinente pour eviter les histogrammes de type "herisson" comportant des
+		// centaines, voire des milliers d'intervalles: cela ameliore vraiment le choix de la granularite par defaut.
+		// Par contre, dans le cas des discretisation avec faible nombre d'intervalles, l'application de cette
+		// regle aboutit parfois a un choix de granulatite trop grossiere, avec perte d'information
+		// De facon heuristique, on choisit donc de ne pas appliquer la regle de Terell-Scott s'il y a moins
+		// de 100 intervalles
+		nMaxPartNumber = max(100, nMaxPartNumber);
+
+		// Nombre max partie en prenant en compte l'eventuelle contrainte utilisateur
 		if (nRequiredMaxPartNumber > 0)
+			nMaxPartNumber = min(nRequiredMaxPartNumber, nMaxPartNumber);
+
+		// On recherche l'histogramme interpretable le plus fin qui respecte la contrainte
+		bestHistogram = NULL;
+		for (n = oaResultHistograms.GetSize() - 1; n >= 0; n--)
 		{
-			// On garde le premier histogramme qui respecte la contrainte
-			if (bestHistogram->GetIntervalNumber() > nRequiredMaxPartNumber)
+			histogram = cast(MHHistogram*, oaResultHistograms.GetAt(n));
```

Comment:
I would need a small comment here as to why this cast is needed, and also whether this is a cast with or without cast control.

Reply:
Pas de commentaire ici, puisque l'on applique des patterns de codage systématique
- cf https://github.com/KhiopsML/khiops/wiki/Coding-Guidelines
- tout classe hérite de la classe Object (cf. [Inheritance](https://github.com/KhiopsML/khiops/wiki/Coding-Guidelines#inheritance), et src/Norm/base/Object.h)
- les seuls containers utilisables sont des containers d'Object (cf. src/Norm/base/Object.h)
  - ObjectArray: tableau
  - ObjectList: liste
  - SortedList; liste triée
  - ObjectDictionary: dictionnaire (table de hashage), avec clé chaine de caractère
  - NumericKeyDictionary: dictionnaire (table de hashage), avec clé numérique
- l'accès à leur élement renvoie des Object, qu'il faut caster, en utilisant a macro `cast`, controlée en mode debug (cf. https://github.com/KhiopsML/khiops/wiki/Coding-Guidelines#programming-by-contract-assertions)

Reply:
OK, I was not aware of the Coding Guidelines, my bad.
I had read the `cast` macro though, but, being unaware of the Guidelines, I had asked for the comment.
But it is Ok and clear now.